File: 17. String/067.py
class Solution:
    def addBinary1(self, a: str, b: str) -> str:
        i = len(a) - 1
        j = len(b) - 1
        res = [0] * (max(i,j)+1)
        carry = 0
        index = 0
        while i >= 0 and j >= 0:
            val = carry + int(a[i]) + int(b[j])
            carry = val // 2
            val = val % 2
            res[index] = val
            i -= 1
            j -= 1
            index += 1

        while i >= 0:
            val = carry + int(a[i])
            carry = val // 2
            val = val % 2
            res[index] = val
            i -= 1
            index += 1

        while j >= 0:
            val = carry + int(b[j])
            carry = val // 2
            val = val % 2
            res[index] = val
            j -= 1
            index += 1

        if carry:
            res.append(1)

        return "".join(map(str,res[::-1]))


    def addBinary(self, a, b):
        s = []
        c = 0
        i = len(a) - 1
        j = len(b) - 1
        while i >= 0 or j >= 0 or c == 1:
            c += int(a[i]) if i >= 0 else 0
            c += int(b[j]) if j >= 0 else 0
            j -= 1
            i -= 1
            s.append(c%2)
            c //= 2

        return "".join(map(str,s[::-1]))

if __name__ == "__main__":
     a = "1010"
     b = "1011"
     print(Solution().addBinary(a,b))

File: C++/067.cpp
#include <string>
#include <iostream>
#include <algorithm>


using namespace std;

/*
        1-11 1-10 1-01 1-00
val     
carry
*/

class Solution {
public:
    string addBinary(string a, string b) {
        unsigned carry = 0;
        string res;
        string::const_reverse_iterator it1{a.crbegin()};
        string::const_reverse_iterator it2{b.crbegin()};

        while(it1 != a.crend() and it2 != b.crend()){
            res.push_back(48 + (*it1 xor *it2 xor carry bitand 1));
            carry = (carry bitand (*it1 bitor *it2) bitand 1) bitor (*it1 bitand *it2 bitand 1);
            ++it1;
            ++it2;
        }

        while(it1 != a.crend()){
            res.push_back(48 + ((*it1 xor carry) bitand 1));
            carry = carry bitand *it1 bitand 1;
            ++it1;
        }
        while(it2 != b.crend()){
            res.push_back(48 + ((*it2 xor carry) bitand 1));
            carry = carry bitand *it2 bitand 1;
            ++it2;
        }

        if(carry == 1){
            res.push_back('1');
        }

        reverse(res.begin(), res.end());
        return res;
    }
};

int main() {
    Solution s;
    cout<<s.addBinary("1", "111")<<endl;
    return 0;
}
